src/runtime/security/client_negotiation.cpp

```diff
@@ -142,6 +142,33 @@ void client_negotiation::on_mechanism_selected(const negotiation_response &resp)
     }
 }
 
+void client_negotiation::on_challenge(const negotiation_response &challenge)
+{
+    if (challenge.status == negotiation_status::type::SASL_CHALLENGE) {
+        std::string response_msg;
+        auto err = _sasl->step(challenge.msg, response_msg);
+        if (!err.is_ok() && err.code() != ERR_SASL_INCOMPLETE) {
+            dwarn_f("{}: negotiation failed, reason = {}", _name, err.description());
+            fail_negotiation();
+            return;
+        }
+
+        auto req = dsn::make_unique<negotiation_request>();
+        _status = req->status = negotiation_status::type::SASL_CHALLENGE_RESP;
+        req->msg = response_msg;
+        send(std::move(req));
```

Comment:
```suggestion
        send(negotiation_status::type::SASL_CHALLENGE_RESP, std::move(response_msg));
```

```
void negotiation::send(negotiation_status::type status, std::string && response_msg) {
        auto req = dsn::make_unique<negotiation_request>();
        _status = req->status = negotiation_status::type::SASL_CHALLENGE_RESP;
        req->msg = std::move(response_msg);
        send(std::move(req));
}

void negotiation::send(negotiation_status::type status) {
        auto req = dsn::make_unique<negotiation_request>();
        req->status = status;
        send(std::move(req));
}
```

You can create some overloaded `send` to reduce the repeated code.

Reply:
OK，thanks.